Onboarding  enabling reactions and apiclient enhancements

weerd opened the pull request:
#### What's this PR do?

This PR updates the `ApiClient` class to fix the `get()` method and add `post()` and `delete()` methods. It also updates how these methods work with `fetch()` and javascript promises, by _correctly_ returning the promise to the location that instantiated the class and called it, so that it may use the `.then()` method to continue the chain when the promise returns a response and allow it to proceed to the next desired functionality.

It also does a bit of code cleanup and adds the `<ReactionsList />` and `<Reaction />` components which work together to output reactions on Reportback Items. The `ReactionList` component uses the updates to the `ApiClient` to both create or delete reactions by the current user on a Reportback Item.

Random, but also trying out leaving in a couple useful `console.log()` calls (and coloring them) with the `ApiClient` so that during local development, we can see what activity is taking place when making API calls. These logs get pulled out during builds on Production!
#### How should this be reviewed?

Pull down code, and checkout out slide 3. Try reacting to a few of the presented RB items. Maybe also doublecheck the database to see if reactions are being created/deleted in the `dosomething_kudos` table as you interact with them.
#### Any background context you want to provide?

Slide 3 now with reactions:
![image](https://cloud.githubusercontent.com/assets/105849/17981660/98065e42-6ad2-11e6-92e5-011fca566e6c.png)

_NOTE:_ One thing to consider... since we're not using React to create the RB items in the Reportback gallery, but we are to create the RB items on Onboarding Slide 3, I could potentially see an issue where a new user signs up for a campaign, sees the onboarding experience and then "likes" a few photos in Slide 3, **_but_** since we're not using React for the gallery, we can't dynamically update RB items in there, if in both areas, the same RB item is presented. This could happen if there are only a few items for a campaign. I've had this happen while testing and for now I don't see an immediate solution. <-- @jessleenyc the "caveat"
#### Relevant tickets

Fixes #6879
Refs #6910
#### Checklist
- [ ] Tested on staging.
- [ ] Pinged a PM if this is a larger PR that would benefit from some additional testing love.

---

@DFurnes @deadlybutter 

cc: @jessleenyc

weerd committed:
Passing user information as a prop.

weerd committed:
Updating and enhancing the ApiClient class with post() and delete() and a bit of cleanup.

weerd committed:
Outputting the ReactionList component in a ReportbackItem.

weerd committed:
Adding the ReactionList and Reaction components and functionality.

weerd commented:
@jessleenyc to resolve the "caveat", we could maybe have a function that gets called when reacting on an RB item in the React onboarding version, that jumps out to find a corresponding item in the gallery, and if so, then change the attributes on it?

itsjoekent commented on lib/themes/dosomething/paraneue_dosomething/js/utilities/ApiClient.js:
```diff
@@ -36,6 +43,37 @@ class ApiClient {
+    const url = `${this.url}${path}`;
+
+    // @TODO: see todo in post() method!
+    console.log('%c ApiClient DELETE Request: \n' + url,
```
whoops?

weerd replied:
Nope. On purpose!

> Random, but also trying out leaving in a couple useful console.log() calls (and coloring them) with the ApiClient so that during local development, we can see what activity is taking place when making API calls. These logs get pulled out during builds on Production!

weerd replied:
This is what it looks like at the moment on local development:

![image](https://cloud.githubusercontent.com/assets/105849/18008396/ea0e9440-6b75-11e6-9691-115c050c0d34.png)

Will help us while developing to know what is getting passed to API calls via the API client 😄

DFurnes replied:
Yep, I think this is a good idea to get in the habit of since there's no downside and can be massively useful for debugging (with useful messages, obvs, not like `console.log('💩')`).

itsjoekent commented on lib/themes/dosomething/paraneue_dosomething/js/utilities/ApiClient.js:
```diff
@@ -36,6 +43,37 @@ class ApiClient {
+      .then(this.checkStatus)
+      .then(this.parseJson)
+      .catch(function(data) {
+        console.log(data);
```
Also should probably remove the .log(data) here & in the one below

itsjoekent replied:
Oh I see in your TODO below

itsjoekent commented:
So im getting a bit lost on the ReactionList component.

We start using the dsKudos data in the settings object,

```
{
 "approved": ["heart"],
 "enabled": true,
 "terms": {"heart":"641"},
 "user":"1700226"}
```

We then go through each approved reaction (approved array) and check if its an available term (terms array) https://github.com/DoSomething/phoenix/pull/6951/files#diff-3d725c871d1c6f13bde530ede939aefeR38

and then we save in the new approved reactions array the value of the term (641)

after we gather this data we set the state with `setInitialReactions`. This goes through every approved reaction we made above & sets the reaction in the state, with the key being the term id

```
    Object.keys(this.approvedReactions).forEach((key) => {
       reactions[key] = {
```

But I get lost here, because right now `key` is always going to be the value of heart, `641`? How do you identify different users reacting...

apparently reaction list is used per reportback item. WHOOPS. 

this really threw me off because (AFAIK?) we're not doing multiple kudo terms, which led me to assume ReactionList was a container of all reactions like the name implies.

DFurnes commented:
> @jessleenyc to resolve the "caveat", we could maybe have a function that gets called when reacting on an RB item in the React onboarding version, that jumps out to find a corresponding item in the gallery, and if so, then change the attributes on it?

I think this is going to be an edge case (especially looking at current kudos engagement). It'd be nice to consider a React refactoring of the gallery later on to tackle some of the technical debt there, rather than layering on a hacky fix for something that shouldn't affect most users.

DFurnes commented on lib/themes/dosomething/paraneue_dosomething/js/components/ReactionsList.js:
```diff
@@ -0,0 +1,158 @@
+
+import Reaction from './Reaction';
+import ApiClient from '../utilities/ApiClient';
+import Setting from '../utilities/Setting';
```
Since this is a function, let's keep it lowercase: `import setting from '../utilities/Setting'`. And maybe we should move it into that `helpers.js` file we made to reduce confusion with the capitalized filename.

weerd replied:
ah, I was wondering why we lowercased it elsewhere. Good call!

DFurnes commented on lib/themes/dosomething/paraneue_dosomething/js/components/ReactionsList.js:
```diff
@@ -0,0 +1,158 @@
+    const approvedReactions = Setting('dsKudosReactions.approved', []);
+    const availableReactions = Setting('dsKudosReactions.terms');
+
+    approvedReactions.forEach((value) => {
```
Could this be simplified down like this? Although it would lose the custom keying, not sure if that's necessary.

``` js
return approvedReactions.filter(value => availableReactions.hasOwnProperty(value));
```

weerd replied:
hmmm, would be cleaner but I'm relying on the custom `key`ing cause I'm relying on it later in `setInitialReactions()`.

DFurnes replied:
Ah, yeah I see. Could maybe do something with Lodash's [keyBy](https://lodash.com/docs#keyBy), but defs not blocking!

weerd replied:
Added it as a `@TODO` to look into using it :)

weerd commented:
@deadlybutter yep yep. It's a little confusing but something we were doing in the `Reportback.js` for the view more gallery too. We need to reconcile between the approved reactions and the available reactions, because there may be like more available but we only want to show one for now.

As for the `key`, yeah, it won't conflict. Say there's a reaction with key `641`, in the ReactionList there can only ever be 1 item with property name `641` that contains the totals and whether current user reacted, _per_ reportback item.

Here's a visual that helps explain. I find pictures do a better job than words:

![reportbackitem reactionlist reactions](https://cloud.githubusercontent.com/assets/105849/18008984/59e5af36-6b78-11e6-8e6d-3c2e5c654ab5.png)

DFurnes commented on lib/themes/dosomething/paraneue_dosomething/js/components/ReactionsList.js:
```diff
@@ -0,0 +1,158 @@
+   */
+  toggleReaction(termId) {
+    const reaction = this.state.reactions[termId];
+    let reactionId = reaction.id;
```
This can be `const` right?

weerd replied:
doh! yep; indeed it can.

itsjoekent commented:
@weerd yeah, originally i was thinking the ReactionList component applied to all the reportbacks, in which case it would have conflicted

DFurnes commented on lib/themes/dosomething/paraneue_dosomething/js/components/ReactionsList.js:
```diff
@@ -0,0 +1,158 @@
+      }).then((response) => {
+        response.forEach((data) => {
+          if (data) {
+            this.updateReaction(termId, data);
```
I'm not sure, but does this cause "n" re-renders because of the `.forEach` and the `this.setState()` in the update function? I forget whether React is smart about batching those up.

itsjoekent replied:
looking at the update function we'd only have 2 re-renders assuming react isnt being smart about it, not a _yuge_ deal

weerd replied:
Hmmm, not sure. Good point tho... is there a good way to check? I didn't see anything crazy going on but also wasn't aware to look for it...

itsjoekent replied:
@weerd I have the React developer tools, adds another tab in the dev console thing

https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en

I think we can track React updates with this, let me check...

itsjoekent replied:
hm, it has a trace updates feature but it just highlights the area which updated. 

I guess a console.log (or console.trace if youre in chrome) might work, who needs fancy shit!

weerd replied:
I've been using the React dev tools too! It's kinda crucial, haha Wasn't sure if there's a better way to check in there is all.

DFurnes replied:
Might be really helpful to include [something like this](https://github.com/ryanflorence/react-lumberjack) in dev builds. (Or as we build more complicated things, it's probably a good argument for using something like Redux)

itsjoekent replied:
that looks neat

DFurnes replied:
For conditionally including in dev build:

``` js
if (process.env.NODE_ENV !== 'production') {
  require('react-lumberjack')
}

```

weerd commented:
@deadlybutter indeed! But good to call it out so we could clarify here! 💃

DFurnes commented:
That drawing is my favorite thing ever. 🎨

weerd commented:
@DFurnes iPad Pro + Apple Pencil + ProCreate. I really want all my PR descriptions to be drawings from now on, haha!

weerd committed:
Updates to address recommendations in PR.
using const instead of let for unchanged variable and using lowercase for 'setting' variable since it is a function and not a class.

DFurnes commented:
🚢

weerd committed:
Update to help with potential multiple unnecessary re-renders of state.

weerd commented:
@DFurnes @deadlybutter new commit addresses avoiding the potential for "n" re-renders. Since it's a Promise I can just chain another [`.then()` method](https://github.com/DoSomething/phoenix/pull/6951/files#diff-3d725c871d1c6f13bde530ede939aefeR135) to update the state only _after_ processing through the `forEach()` loop 💃 Seems to work fine!

weerd committed:
Reverting back to prior approach after IRL convo with @dfurnes.

weerd commented:
Nevermind. Reverted back after IRL convo w/ @DFurnes!

itsjoekent commented:
@weerd why did you revert?

weerd commented:
Talked it over with @DFurnes and it's really just not a good idea to have the state mutations happening separate from the actual setting of the state via React API; it could lead to issues and bugs. We talked it over and how I had it before should be fine, but eventually we should consider making changes in a cloned object and then passing that into the state via React API. Ideally, we'd also move to using something like Redux...
